modules/nodesale/purchase.go

```diff
@@ -0,0 +1,101 @@
+package nodesale
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/cockroachdb/errors"
+	"github.com/gaze-network/indexer-network/core/types"
+	"github.com/gaze-network/indexer-network/modules/nodesale/datagateway"
+	purchasevalidator "github.com/gaze-network/indexer-network/modules/nodesale/internal/validator/purchase"
+	"github.com/gaze-network/indexer-network/pkg/logger"
+	"github.com/gaze-network/indexer-network/pkg/logger/slogx"
+)
+
+func (p *Processor) processPurchase(ctx context.Context, qtx datagateway.NodesaleDataGatewayWithTx, block *types.Block, event nodesaleEvent) error {
+	purchase := event.eventMessage.Purchase
+	payload := purchase.Payload
+
+	validator := purchasevalidator.New()
+
+	_, err := validator.EqualXonlyPublicKey(payload.BuyerPublicKey, event.txPubkey)
+	if err != nil {
+		logger.DebugContext(ctx, "Invalid public key", slogx.Error(err))
+	}
+
+	_, deploy, err := validator.NodeSaleExists(ctx, qtx, payload)
+	if err != nil {
+		return errors.Wrap(err, "Cannot connect to datagateway")
+	}
+
+	validator.ValidTimestamp(deploy, block.Header.Timestamp)
+	validator.WithinTimeoutBlock(payload, uint64(event.transaction.BlockHeight))
```

Comment:
Should pick one between
1. All validator funcs return an error on condition fails OR
2. All validator funcs don't return errors, and stores the valid bool and error inside the validator, which can be read later when we want to know the error.